### ISSUE TITLE
Assertion failure when a function takes a variadic function

### ISSUE THREAD

**Issue by m-rots** [apple/swift#65368]
**Description**
The following assertion failure is raised upon compiling a function that takes a variadic function:

```
Assertion failed: (param.isPack() == isa<SILPackType>(param.getInterfaceType()) && "Packs must have pack convention")
```

**Reproduction**

The following reproduction has been found inside [`test/SILGen/variadic-generic-reabstraction.swift`](https://github.com/apple/swift/blob/1c3667b52aa7888f7e2cc253b725a7ed055e4ef2/test/SILGen/variadic-generic-reabstraction.swift#L3):

```swift
func takesVariadicFunction<each T>(_ function: (repeat each T) -> Int) {}
```

**Environment**
- Swift compiler version info: `Apple Swift version 5.9-dev (LLVM 107de6a91bc1dbd, Swift 022311e438e7703)`
- Xcode version info: `Xcode 14.3 Build version 14E222b`
- Deployment target: `arm64-apple-macosx13.0`

**Stacktrace**
```
Assertion failed: (param.isPack() == isa<SILPackType>(param.getInterfaceType()) && "Packs must have pack convention"), function SILFunctionType, file ASTContext.cpp, line 4530.

Stack dump:
0.	Program arguments: /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2023-04-20-a.xctoolchain/usr/bin/swift-frontend -frontend -c -primary-file /Users/storm/Developer/Swift/Repro/Sources/Repro/Repro.swift -emit-dependencies-path /Users/storm/Developer/Swift/Repro/.build/arm64-apple-macosx/debug/Repro.build/Repro.d -emit-reference-dependencies-path /Users/storm/Developer/Swift/Repro/.build/arm64-apple-macosx/debug/Repro.build/Repro.swiftdeps -target arm64-apple-macosx10.13 -Xllvm -aarch64-use-tbi -enable-objc-interop -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -I /Users/storm/Developer/Swift/Repro/.build/arm64-apple-macosx/debug -I /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib -I /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/usr/lib -F /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -F /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -color-diagnostics -enable-testing -g -module-cache-path /Users/storm/Developer/Swift/Repro/.build/arm64-apple-macosx/debug/ModuleCache -swift-version 5 -Onone -D SWIFT_PACKAGE -D DEBUG -new-driver-path /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2023-04-20-a.xctoolchain/usr/bin/swift-driver -empty-abi-descriptor -plugin-path /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2023-04-20-a.xctoolchain/usr/lib/swift/host/plugins -plugin-path /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2023-04-20-a.xctoolchain/usr/local/lib/swift/host/plugins -resource-dir /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2023-04-20-a.xctoolchain/usr/lib/swift -enable-anonymous-context-mangled-names -Xcc -isysroot -Xcc /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -Xcc -F -Xcc /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks -Xcc -fPIC -module-name Repro -target-sdk-version 13.3 -target-sdk-name macosx13.3 -parse-as-library -o /Users/storm/Developer/Swift/Repro/.build/arm64-apple-macosx/debug/Repro.build/Repro.swift.o -index-store-path /Users/storm/Developer/Swift/Repro/.build/arm64-apple-macosx/debug/index/store -index-system-modules
1.	Apple Swift version 5.9-dev (LLVM 107de6a91bc1dbd, Swift 022311e438e7703)
2.	Compiling with the current language version
3.	While evaluating request IRGenRequest(IR Generation for file "/Users/storm/Developer/Swift/Repro/Sources/Repro/Repro.swift")
4.	While emitting IR SIL function "@$s5Repro21takesVariadicFunctionyySixxQpXERvzlF".
 for 'takesVariadicFunction(_:)' (at /Users/storm/Developer/Swift/Repro/Sources/Repro/Repro.swift:1:1)
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  swift-frontend           0x00000001075f753c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  swift-frontend           0x00000001075f68e0 llvm::sys::RunSignalHandlers() + 112
2  swift-frontend           0x00000001075f7b7c SignalHandler(int) + 304
3  libsystem_platform.dylib 0x0000000184d8aa84 _sigtramp + 56
4  libsystem_pthread.dylib  0x0000000184d5bc28 pthread_kill + 288
5  libsystem_c.dylib        0x0000000184c69ae8 abort + 180
6  libsystem_c.dylib        0x0000000184c68e44 err + 0
7  swift-frontend           0x0000000107a5679c swift::SILFunctionType::SILFunctionType(swift::GenericSignature, swift::SILExtInfo, swift::SILCoroutineKind, swift::ParameterConvention, llvm::ArrayRef<swift::SILParameterInfo>, llvm::ArrayRef<swift::SILYieldInfo>, llvm::ArrayRef<swift::SILResultInfo>, llvm::Optional<swift::SILResultInfo>, swift::SubstitutionMap, swift::SubstitutionMap, swift::ASTContext const&, swift::RecursiveTypeProperties, swift::ProtocolConformanceRef) (.cold.11) + 0
8  swift-frontend           0x0000000103c3ce98 swift::SILFunctionType::SILFunctionType(swift::GenericSignature, swift::SILExtInfo, swift::SILCoroutineKind, swift::ParameterConvention, llvm::ArrayRef<swift::SILParameterInfo>, llvm::ArrayRef<swift::SILYieldInfo>, llvm::ArrayRef<swift::SILResultInfo>, llvm::Optional<swift::SILResultInfo>, swift::SubstitutionMap, swift::SubstitutionMap, swift::ASTContext const&, swift::RecursiveTypeProperties, swift::ProtocolConformanceRef) + 1792
9  swift-frontend           0x0000000103c3d8d8 swift::SILFunctionType::get(swift::GenericSignature, swift::SILExtInfo, swift::SILCoroutineKind, swift::ParameterConvention, llvm::ArrayRef<swift::SILParameterInfo>, llvm::ArrayRef<swift::SILYieldInfo>, llvm::ArrayRef<swift::SILResultInfo>, llvm::Optional<swift::SILResultInfo>, swift::SubstitutionMap, swift::SubstitutionMap, swift::ASTContext const&, swift::ProtocolConformanceRef) + 1184
10 swift-frontend           0x0000000103c6c210 swift::Demangle::ASTBuilder::createImplFunctionType(swift::Demangle::ImplParameterConvention, llvm::ArrayRef<swift::Demangle::ImplFunctionParam<swift::Type>>, llvm::ArrayRef<swift::Demangle::ImplFunctionResult<swift::Type>>, llvm::Optional<swift::Demangle::ImplFunctionResult<swift::Type>>, swift::Demangle::ImplFunctionTypeFlags) + 936
11 swift-frontend           0x0000000103c702ec swift::Demangle::TypeDecoder<swift::Demangle::ASTBuilder>::decodeMangledType(swift::Demangle::Node*, unsigned int, bool) + 8956
12 swift-frontend           0x0000000103c6a374 swift::Demangle::getTypeForMangling(swift::ASTContext&, llvm::StringRef, swift::GenericSignature) + 152
13 swift-frontend           0x0000000102df7c88 (anonymous namespace)::IRGenDebugInfoImpl::getOrCreateType(swift::irgen::DebugTypeInfo) + 1260
14 swift-frontend           0x0000000102df9060 (anonymous namespace)::IRGenDebugInfoImpl::createParameterType(llvm::SmallVectorImpl<llvm::Metadata*>&, swift::SILType) + 104
15 swift-frontend           0x0000000102df8d8c (anonymous namespace)::IRGenDebugInfoImpl::createParameterTypes(swift::CanTypeWrapper<swift::SILFunctionType>) + 400
16 swift-frontend           0x0000000102df1cec (anonymous namespace)::IRGenDebugInfoImpl::emitFunction(swift::SILDebugScope const*, llvm::Function*, swift::SILFunctionTypeRepresentation, swift::SILType, swift::DeclContext*, llvm::StringRef) + 1632
17 swift-frontend           0x0000000102df264c (anonymous namespace)::IRGenDebugInfoImpl::emitFunction(swift::SILFunction&, llvm::Function*) + 160
18 swift-frontend           0x0000000102e2d100 (anonymous namespace)::IRGenSILFunction::emitSILFunction() + 892
19 swift-frontend           0x0000000102e2c824 swift::irgen::IRGenModule::emitSILFunction(swift::SILFunction*) + 948
20 swift-frontend           0x0000000102cd6f68 swift::irgen::IRGenerator::emitGlobalTopLevel(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) + 684
21 swift-frontend           0x0000000102ddc7fc swift::IRGenRequest::evaluate(swift::Evaluator&, swift::IRGenDescriptor) const + 1620
22 swift-frontend           0x0000000102e2bf48 swift::GeneratedModule swift::SimpleRequest<swift::IRGenRequest, swift::GeneratedModule (swift::IRGenDescriptor), (swift::RequestFlags)9>::callDerived<0ul>(swift::Evaluator&, std::__1::integer_sequence<unsigned long, 0ul>) const + 208
23 swift-frontend           0x0000000102debabc llvm::Expected<swift::IRGenRequest::OutputType> swift::Evaluator::getResultUncached<swift::IRGenRequest>(swift::IRGenRequest const&) + 364
24 swift-frontend           0x0000000102ddeb44 swift::performIRGeneration(swift::FileUnit*, swift::IRGenOptions const&, swift::TBDGenOptions const&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule>>, llvm::StringRef, swift::PrimarySpecificPaths const&, llvm::StringRef, llvm::GlobalVariable**) + 212
25 swift-frontend           0x00000001029cd5c0 performCompileStepsPostSILGen(swift::CompilerInstance&, std::__1::unique_ptr<swift::SILModule, std::__1::default_delete<swift::SILModule>>, llvm::PointerUnion<swift::ModuleDecl*, swift::SourceFile*>, swift::PrimarySpecificPaths const&, int&, swift::FrontendObserver*) + 1744
26 swift-frontend           0x00000001029cc924 swift::performCompileStepsPostSema(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 568
27 swift-frontend           0x00000001029dc384 withSemanticAnalysis(swift::CompilerInstance&, swift::FrontendObserver*, llvm::function_ref<bool (swift::CompilerInstance&)>, bool) + 160
28 swift-frontend           0x00000001029cf364 performCompile(swift::CompilerInstance&, int&, swift::FrontendObserver*) + 620
29 swift-frontend           0x00000001029ce370 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 2416
30 swift-frontend           0x00000001028551ec swift::mainEntry(int, char const**) + 3216
31 dyld                     0x0000000184a03f28 start + 2236
```

**Comment by tevelee** [comment]
The same thing happened to me. Here's a short example project to demonstrate the issue: https://github.com/tevelee/VariadicGenerics

Specifically, adding [this](https://github.com/tevelee/VariadicGenerics/blob/main/Sources/VariadicGenerics/VariadicGenerics.swift#L26-L28) function:
```swift
extension ServiceResolver {
    func instance<T, each P>(from initializer: (repeat each P) -> T) -> T {
        initializer(repeat resolve() as each P)
    }
}
```

**Comment by tevelee** [comment]
Update: this no longer fails for me with Xcode 15 beta 1 and Swift 5.9

**Comment by AnthonyLatsis** [comment]
@slavapestov Do we know what fixed this?

**Comment by slavapestov** [comment]
@AnthonyLatsis I believe it was this:

```
commit 1c924018d1f0f8ba41161f96d1b353d2f361dfc0 (slava/sil-function-type-pack-mangling)
Author: Slava Pestov <spestov@apple.com>
Date:   Mon May 1 18:14:37 2023 -0400

    ASTMangler: Fix mangling of SILFunctionType with @pack conventions
    
    Fixes rdar://problem/107151125.
```

**Comment by AnthonyLatsis** [comment]
Thanks!
